File: CombineDemo/simple_server.py
from http.server import BaseHTTPRequestHandler, HTTPServer
import json
from urllib.parse import urlparse, parse_qs

# 사용 불가능한 사용자 이름 목록
unavailable_usernames = ['jmbae', 'johnnyappleseed', 'page', 'johndoe']

class UserNameHandler(BaseHTTPRequestHandler):
    def do_GET(self):
        # URL 파싱
        parsed_url = urlparse(self.path)
        
        # 요청 경로가 /isUserNameAvailable인지 확인
        if parsed_url.path == '/isUserNameAvailable':
            # 쿼리 파라미터 파싱
            query_params = parse_qs(parsed_url.query)
            
            # userName 파라미터가 있는지 확인
            if 'userName' in query_params:
                username = query_params['userName'][0]
                # 금지된 사용자 이름 확인
                if username in ['admin', 'superuser']:
                    self.send_response(400)
                    self.send_header('Content-Type', 'application/json')
                    self.end_headers()
                    error_response = {
                        "error": True,
                        "reason": f"Username is not valid: {username}."
                    }
                    self.wfile.write(json.dumps(error_response).encode())
                    return
                
                # 서버 에러 시뮬레이션 - 데이터베이스 손상
                if username == 'servererror':
                    self.send_response(500)
                    self.send_header('Content-Type', 'application/json')
                    self.end_headers()
                    error_response = {
                        "error": True,
                        "reason": "The database is corrupted"
                    }
                    self.wfile.write(json.dumps(error_response).encode())
                    return
                
                # 유지보수 에러 시뮬레이션
                if not hasattr(self.__class__, 'maintenance_counter'):
                    self.__class__.maintenance_counter = 0
                
                if username == 'maintenance':
                    self.__class__.maintenance_counter += 1
                    print(f"Maintenance counter: {self.__class__.maintenance_counter}")
                    
                    if self.__class__.maintenance_counter % 3 != 0:
                        print("... throwing maintenance error")
                        self.send_response(500)
                        self.send_header('Content-Type', 'application/json')
                        self.send_header('Retry-After', '120')
                        self.end_headers()
                        error_response = {
                            "error": True,
                            "reason": "Temporarily unavailable for maintenance"
                        }
                        self.wfile.write(json.dumps(error_response).encode())
                        return
                    else:
                        print("... NOT throwing maintenance error")
                
                # 항상 유지보수 에러
                if username == 'maintenance!':
                    self.send_response(500)
                    self.send_header('Content-Type', 'application/json')
                    self.send_header('Retry-After', '120')
                    self.end_headers()
                    error_response = {
                        "error": "Internal Server Error",
                        "reason": "Temporarily unavailable for maintenance"
                    }
                    self.wfile.write(json.dumps(error_response).encode())
                    return
                
                # 잘못된 응답 형식 시뮬레이션
                if username == 'illegalresponse':
                    self.send_response(200)
                    self.send_header('Content-Type', 'application/json')
                    self.end_headers()
                    result = {"isAvailable": False}
                    self.wfile.write(json.dumps(result).encode())
                    return
                    
                # 사용자 이름이 비어있는지 확인
                if not username:
                    self.send_response(400)
                    self.end_headers()
                    self.wfile.write(b'Bad Request: userName 이 비어있습니다.')
                    return

                # 사용자 이름이 3자 이상인지 확인
                if len(username) < 3:
                    self.send_response(400)
                    self.end_headers()
                    self.wfile.write(b'Bad Request: userName은 최소 3자 이상이어야 합니다.')
                    return
                
                # 요청 로깅
                print(f"Checking availability for username: {username}")
                
                # 사용자 이름이 사용 불가능한 목록에 있는지 확인
                is_available = username not in unavailable_usernames
                
                # JSON 응답 생성
                response = {
                    "isAvailable": is_available,
                    "userName": username
                }
                
                # HTTP 응답 헤더 설정
                self.send_response(200)
                self.send_header('Content-Type', 'application/json')
                self.end_headers()
                
                # JSON 응답 전송
                self.wfile.write(json.dumps(response).encode())
                return
        
        # 요청 경로가 일치하지 않는 경우 404 반환
        self.send_response(404)
        self.end_headers()
        self.wfile.write(b'Not Found')

def run_server(port=8080):
    server_address = ('127.0.0.1', port)
    httpd = HTTPServer(server_address, UserNameHandler)
    print(f"Starting username availability server at http://127.0.0.1:{port}")
    print("Available endpoints:")
    print(f"  GET http://127.0.0.1:{port}/isUserNameAvailable?userName=<username>")
    print("\nUnavailable usernames for testing:")
    for name in unavailable_usernames:
        print(f"  - {name}")
    print("\n서버를 중지하려면 Ctrl+C를 누르세요.")
    
    try:
        httpd.serve_forever()
    except KeyboardInterrupt:
        print("\n서버가 종료됩니다.")
        httpd.server_close()

if __name__ == '__main__':
    run_server()

File: CombineDemo/simple_server.rb
require 'socket'
require 'json'
require 'uri'

# 사용 불가능한 사용자 이름 목록
UNAVAILABLE_USERNAMES = ['jmbae', 'johnnyappleseed', 'page', 'johndoe']


# HTTP 응답 헬퍼 메서드
def http_response(code, content_type, body)
  "HTTP/1.1 #{code}\r\n" +
  "Content-Type: #{content_type}\r\n" +
  "Content-Length: #{body.bytesize}\r\n" +
  "Connection: close\r\n" +
  "\r\n" +
  body
end

# 쿼리 파라미터 파싱 헬퍼 메서드
def parse_query_params(query_string)
  params = {}
  return params if query_string.nil? || query_string.empty?
  
  query_string.split('&').each do |param|
    key, value = param.split('=', 2)
    params[URI.decode_www_form_component(key)] = URI.decode_www_form_component(value || '')
  end
  
  params
end

# 서버 실행 함수
def run_server(port=8080)
  maintenance_counter = 0
  # 서버 소켓 생성
  server = TCPServer.new('127.0.0.1', port)
  
  # 시작 메시지 출력
  puts "Starting username availability server at http://127.0.0.1:#{port}"
  puts "Available endpoints:"
  puts "  GET http://127.0.0.1:#{port}/isUserNameAvailable?userName=<username>"
  puts "\nUnavailable usernames for testing:"
  UNAVAILABLE_USERNAMES.each do |name|
    puts "  - #{name}"
  end
  puts "\n서버를 중지하려면 Ctrl+C를 누르세요."
  
  # 인터럽트 시그널(Ctrl+C) 처리
  running = true
  trap('INT') do
    if running
      running = false
      puts "\n서버가 종료됩니다."
      server.close rescue nil
      exit
    else
      # 두 번째 Ctrl+C가 눌릴 경우 강제 종료
      puts "\n강제 종료합니다."
      exit!
    end
  end
  
  # 클라이언트 요청 처리 루프
  loop do
    # 클라이언트 연결 수락
    client = nil
    
    begin
      # 클라이언트 연결 수락 - 타임아웃 설정
      client = server.accept_nonblock
    rescue IO::WaitReadable, Errno::EINTR
      # 논블로킹 모드에서 소켓이 준비되지 않았을 때 발생하는 예외 처리
      IO.select([server])
      retry
    end
    
    # 각 클라이언트 요청을 별도 스레드에서 처리
    Thread.new(client) do |c|
      begin
        # 시간 초과 설정
        c.setsockopt(Socket::SOL_SOCKET, Socket::SO_RCVTIMEO, [5, 0].pack('l_*'))
        c.setsockopt(Socket::SOL_SOCKET, Socket::SO_SNDTIMEO, [5, 0].pack('l_*'))
        
        # 요청 데이터 읽기
        request_line = c.gets
        
        # nil 체크 - 빈 요청인 경우
        if request_line.nil? || request_line.strip.empty?
          c.close
          next
        end
        
        # 요청 정보 파싱
        parts = request_line.split
        
        # 유효한 HTTP 요청인지 확인
        if parts.length < 2
          c.close
          next
        end
        
        method, path_with_query = parts[0], parts[1]
        
        # 경로와 쿼리 문자열 분리
        if path_with_query.include?('?')
          path, query_string = path_with_query.split('?', 2)
        else
          path = path_with_query
          query_string = ''
        end
        
        # 요청 헤더 건너뛰기
        while line = c.gets
          break if line.strip.empty?
        end
        
        # 요청 처리
        if path == '/isUserNameAvailable' && method == 'GET'
          params = parse_query_params(query_string)
          
          if params.has_key?('userName')
            username = params['userName']

            if username.nil? || username.empty?
              # 사용자 이름이 비어 있는 경우
              error_response = {
                error: true,
                reason: 'userName을 입력하세요'
              }
              response = http_response(
                '400 Bad Request',
                'application/json',
                JSON.generate(error_response)
              )
              c.write(response)
              c.close
              next
            end

            # 3자 이상인지 확인
            if username.length < 3
              # 사용자 이름이 3자 미만인 경우
              error_response = {
                error: true,
                reason: 'userName은 3자 이상이어야 합니다'
              }
  
              response = http_response(
                '400 Bad Request',
                'application/json',
                JSON.generate(error_response)
              )
              c.write(response)
              c.close
              next
            end

            # 금지된 사용자 이름 확인
            if ['admin', 'superuser'].include?(username)
              error_response = {
                error: true,
                reason: "Username is not valid: #{username}."
              }
              response = http_response(
                '400 Bad Request',
                'application/json',
                JSON.generate(error_response)
              )
              c.write(response)
              c.close
              next
            end

            # 서버 에러 시뮬레이션 - 데이터베이스 손상
            if username == 'servererror'
              error_response = {
                error: true,
                reason: 'The database is corrupted'
              }
              response = http_response(
                '500 Internal Server Error',
                'application/json',
                JSON.generate(error_response)
              )
              c.write(response)
              c.close
              next
            end

            # 유지보수 에러 시뮬레이션            
            if username == 'maintenance'
              maintenance_counter += 1
              puts "Maintenance counter: #{maintenance_counter}"
              
              if maintenance_counter % 3 != 0
                puts "... throwing maintenance error"
                error_response = {
                  error: true,
                  reason: 'Temporarily unavailable for maintenance'
                }
                response = http_response(
                  '500 Internal Server Error',
                  'application/json',
                  JSON.generate(error_response)
                )
                # 재시도 헤더 추가
                response = response.sub("Connection: close\r\n", "Connection: close\r\nRetry-After: 12\r\n")
                c.write(response)
                c.close
                next
              else
                puts "... NOT throwing maintenance error"
              end
            end

            # 항상 유지보수 에러
            if username == 'maintenance!'
              error_response = {
                error: 'Internal Server Error',
                reason: 'Temporarily unavailable for maintenance'
              }
              response = http_response(
                '500 Internal Server Error',
                'application/json',
                JSON.generate(error_response)
              )
              # 재시도 헤더 추가
              response = response.sub("Connection: close\r\n", "Connection: close\r\nRetry-After: 12\r\n")
              c.write(response)
              c.close
              next
            end

            # 잘못된 응답 형식 시뮬레이션
            if username == 'illegalresponse'
              result = { isAvailable: false }
              response = http_response(
                '200 OK',
                'application/json',
                JSON.generate(result)
              )
              c.write(response)
              c.close
              next
            end
            
            # 요청 로깅
            puts "Checking availability for username: #{username}"
            
            # 사용자 이름이 사용 불가능한 목록에 있는지 확인
            is_available = !UNAVAILABLE_USERNAMES.include?(username)
            
            # JSON 응답 생성
            result = {
              isAvailable: is_available,
              userName: username
            }
            
            # JSON 응답 전송
            response = http_response(
              '200 OK',
              'application/json',
              JSON.generate(result)
            )
            c.write(response)
          else
            # userName 파라미터가 없는 경우
            response = http_response(
              '400 Bad Request',
              'text/plain',
              'Bad Request: userName parameter is required'
            )
            c.write(response)
          end
        else
          # 요청 경로가 일치하지 않는 경우
          response = http_response(
            '404 Not Found',
            'text/plain',
            'Not Found'
          )
          c.write(response)
        end
      rescue => e
        puts "Error handling request: #{e.message}"
      ensure
        # 항상 클라이언트 연결을 닫음
        c.close rescue nil
      end
    end
  end
end

# 메인 실행
if __FILE__ == $0
  run_server
end